scipio/src/controllers/deadline_queue.rs

```diff
@@ -0,0 +1,470 @@
+// Unless explicitly stated otherwise all files in this repository are licensed under the
+// MIT/Apache-2.0 License, at your convenience
+//
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2020 Datadog, Inc.
+
+use crate::channels::{LocalChannel, LocalReceiver, LocalSender};
+use crate::collections::Deque;
+use crate::{enclose, task};
+use crate::{Latency, Local, Shares, SharesManager, TaskQueueHandle};
+use core::fmt::Debug;
+use futures::stream::StreamExt;
+use std::cell::{Cell, RefCell};
+use std::collections::VecDeque;
+use std::future::Future;
+use std::io;
+use std::pin::Pin;
+use std::rc::Rc;
+use std::time::{Duration, Instant};
+
+/// Items going into the [`DeadlineQueue`] must implement this trait.
+///
+/// It allows the [`DeadlineQueue`] to understand the progress and expectations
+/// of processing this item
+///
+/// [`DeadlineQueue`]: struct.DeadlineQueue.html
+pub trait DeadlineSource {
+    /// What type is returned by the [`action`] method
+    ///
+    /// [`action`]: trait.DeadlineSource.html#method.action
+    type Output;
+
+    /// Returns a [`Duration`] indicating when we would like this operation to complete.
+    ///
+    /// It is calculated from the point of Queueing, not from the point in which the operation
+    /// starts.
+    fn deadline(&self) -> Duration;
```

Comment:
Without reading the code closely, the signature feels weird. I'd expect `fn deadline(&self) -> Instant`

Reply:
right, so maybe the name needs to be changed.
This represents how long the operation is supposed to run for - we use that to calculate speed later so this has to be a duration.

When we enqueue it we add that to now() to obtain an instant, which is then placed in the queue collection.

Maybe just call it `duration` ?

Reply:
Called it `expected_duration`